plugins/in_mem_metrics/mem_metrics.c

```diff
@@ -0,0 +1,458 @@
+    ctx->a = cmt_gauge_create(ctx->cmt, "node", "smaps_rollup", #a, (b), \
+                                  2, (char *[]) {"pid", "type"}); \
+    if (ctx->a == NULL) { \
+        flb_plg_error(ctx->ins, "unable to allocate gauge: "#a); \
+        goto gauge_error; \
+    }
+
+/* Initialize plugin */
+static int cb_mem_metrics_init(struct flb_input_instance *ins,
+                               struct flb_config *config, void *data)
+{
+    int ret;
+    struct mem_metrics *ctx = NULL;
+
+    ctx = flb_calloc(1, sizeof(struct mem_metrics));
+    if (!ctx) {
+        flb_errno();
+        return -1;
+    }
+
+    ret = flb_input_config_map_set(ins, (void *) ctx);
+    if (ret == -1) {
+        return -1;
+    }
+    ctx->ins = ins;
+
+    flb_input_set_context(ins, ctx);
+
+    ctx->cmt = cmt_create();
+
+    MEM_METRICS_GAUGE_INIT(rss, "RSS");
+    MEM_METRICS_GAUGE_INIT_TYPED(pss, "PSS");
+    MEM_METRICS_GAUGE_INIT_TYPED(shared, "Shared Memory");
+    MEM_METRICS_GAUGE_INIT_TYPED(private, "Private Memory");
+    MEM_METRICS_GAUGE_INIT(referenced, "Referenced Memory");
+    MEM_METRICS_GAUGE_INIT(anonymous, "Anonymous Memory");
+    MEM_METRICS_GAUGE_INIT(lazy_free, "Lazy Free Memory");
+    MEM_METRICS_GAUGE_INIT(anon_huge_pages, "Anonymous Huge Pages");
+    MEM_METRICS_GAUGE_INIT(shmem_pmd_mapped, "Shared PMD Mapped");
+    MEM_METRICS_GAUGE_INIT(file_pmd_mapped, "File PMD Mapped");
+    MEM_METRICS_GAUGE_INIT(shared_hugetlb, "Shared HugeTLB");
+    MEM_METRICS_GAUGE_INIT(private_hugetlb, "Private HugeTLB");
+    MEM_METRICS_GAUGE_INIT(swap, "Swap Memory");
+    MEM_METRICS_GAUGE_INIT(swap_pss, "Swap PSS Memory");
+    MEM_METRICS_GAUGE_INIT(locked, "Locked Memory");
+
+    /* unit test 0: collector_time */
+    ret = flb_input_set_collector_time(ins, cb_collector_time,
+                                       ctx->interval_sec, ctx->interval_nsec,
+                                       config);
+    if (ret < 0) {
+        cmt_destroy(ctx->cmt);
+        flb_free(ctx);
+        return -1;
+    }
+    ctx->coll_fd = ret;
+
+    return 0;
+gauge_error:
+    cmt_destroy(ctx->cmt);
+    flb_free(ctx);
+    return -1;
+}
+
+static int cb_mem_metrics_exit(void *data, struct flb_config *config)
+{
+    struct mem_metrics *ctx = data;
+
+    cmt_destroy(ctx->cmt);
+    flb_free(ctx);
+    return 0;
+}
+
+/* Configuration properties map */
+static struct flb_config_map config_map[] = {
+    {
+     FLB_CONFIG_MAP_STR, "proc_path", "/proc",
+     0, FLB_TRUE, offsetof(struct mem_metrics, procfs_path),
+     "The path of the proc pseudo filesystem, default: /proc"
+    },
+    {
+     FLB_CONFIG_MAP_STR, "filter_exec", NULL,
+     0, FLB_TRUE, offsetof(struct mem_metrics, chosen_exec),
+     "Filter for a single executable"
+    },
+    {
+     FLB_CONFIG_MAP_STR, "filter_cmd", NULL,
+     0, FLB_TRUE, offsetof(struct mem_metrics, chosen_cmd),
+     "Filter by the command line"
+    },
+    {
+      FLB_CONFIG_MAP_INT, "filter_pid", "-1",
```

Comment:
I'm thinking of adding an option to add cgroup information when requested.

I'll also make sure to add documentation for using it with k8s. I'm thinking to do so you'd need to add a node name label and add a hostPath mount to k8s for the pod in a daemonset and point the plugin's procfs to that mounted directory.

Reply:
And yeah, multiple PIDs in a list would be nice.

> Dropping the other filters would cripple the functionality way too much in my opinion though, especially since the cmdline filter is the one I usually use.

I have no desire to remove any of the other filters, just add multiple value support for filtering.

Reply:
I have just pushed code that should allow for comma delimited values for any of the filters, including PIDs (with support for the special values self and 0). 

Adding support for negative values (similar to how wait(1) works) for tracking process groups might also be useful.